### ISSUE TITLE
Adding a header variant of toCsv

### ISSUE THREAD

**Issue by mcgizzle** [domdere/cassava-conduit#28]
Thanks for this package, its super useful.

I want to add a new method to allow streaming out to a CSV file using a header, but I am strugglig to build the project with Mafia.

I think the code should be pretty simple:

```
toNamedCsv :: (Monad m, ToNamedRecord a) => Header -> EncodeOptions -> ConduitT a BS.ByteString m ()
toNamedCsv header opts = {-# SCC toCsv_p #-} map $ BSL.toStrict . encodeWithName header opts . pure
```

Anyway, let me know if something like this would be accepted and I'll persevere setting up mafia 😅

**Comment by domdere** [comment]
Hmm yeah that would probably be useful.

Feel free to remove mafia from the project and make it rely on vanilla
cabal (just not `stack` please), it shouldn't require much more than adding
a cabal lock file.

That would make it more friendly for others

**Comment by domdere** [comment]
So a lot of people do ask for this one, i've never added because i dont really see the point, the header is just one line, usually when i do this i write the header line, flush it and begin streaming the rows.

I think the code you have provided might actually create a header for each row? In that case I don't think it would have a wide use case.

If it were done properly, I'd probably be ok with adding it, since a lot of people do ask for it for some reason.

But TBH i think the Conduit code for it would be more trouble than its worth, you would have to make sure it only outputs the header line before the first row (which i think you can do by concating two streams, where the first is just a "stream" with one row in it, the header line), when its much simpler to do this in plain `IO` which is more for stuff of the "run this instruction once" flavour, while `Conduit` is more for stuff of "on every chunk of input do these instructions" flavour...

**Comment by mcgizzle** [comment]
> usually when i do this i write the header line, flush it and begin streaming the rows.

do you have some example code of how this is done?

I think you're probably right that its unnesecary if its not going to be as straigtforward as I had hoped.

**Comment by domdere** [comment]
um yeah sure?

the sample provided in the README does something like this:

``` haskell
-- |
--  A Conduit pipeline that streams from '../exampledata/sampleinput.psv', decodes it from a pipe seperated format,
--  processes it with 'processInput' and the encodes it to pipe seperated format and streams it out to '../exampledata/sampleoutput.psv'
--  The first time it encounters a parse error, it will stop streaming and return the error, dropping any decoded records that came through in that batch also...
--
conduitPipeline :: (MonadError CsvParseError m, MonadResource m) => m ()
conduitPipeline = sourceFile "../exampledata/sampleinput.psv" $$ fromCsv (decodeOpts $ fromIntegral $ ord '|') HasHeader =$= map processInput =$= toCsv (encodeOpts $ fromIntegral $ ord '|') =$= sinkFile "../exampledata/sampleoutput.psv"

main :: IO ()
main = do
    res <- runEitherT $ bimapEitherT showError id $ runResourceT conduitPipeline
    either putStrLn return res
```

To write the header first you would  change it to first open a handle and use that in place of the filename:

``` haskell
import System.IO
import Data.Text qualified as T
import Data.Text.IO as T
-- |
--  A Conduit pipeline that streams from '../exampledata/sampleinput.psv', decodes it from a pipe seperated format,
--  processes it with 'processInput' and the encodes it to pipe seperated format and streams it out to '../exampledata/sampleoutput.psv'
--  The first time it encounters a parse error, it will stop streaming and return the error, dropping any decoded records that came through in that batch also...
--
conduitPipeline :: (MonadError CsvParseError m, MonadResource m) => Handle -> m ()
conduitPipeline h = sourceFile "../exampledata/sampleinput.psv" $$ fromCsv (decodeOpts $ fromIntegral $ ord '|') HasHeader =$= map processInput =$= toCsv (encodeOpts $ fromIntegral $ ord '|') =$= sinkHandle h

main :: IO ()
main = withHandle "../exampledata/sampleoutput.psv" WriteMode $ \h -> do
    -- write the header to the output
    T.hPutStrLn h "Column1Title,Column2Title" -- You can also use Data.ByteString if you want to be more deliberate about your locales and encodings, explicitly converting from Text to a ByteString, Im just using Data.Text here for simplicity.
    -- after writing the one off header,
    -- open up the input file and stream it in, processing in finite chunks at a time to save memory
    res <- runEitherT $ bimapEitherT showError id $ runResourceT (conduitPipeline h)
    either putStrLn return res
```

I haven't really had time to check if this code actually compiles, sorry, its just pseudo code...

However you should be able to google and find many resources on writing text to files and using file handled in Haskell.

You'll notice it doesn't use `ToNamedRecord` to get the header line, sorry if this disappoints.

`ToNamedRecord` is not a great interface however, observe the example instance below:

``` haskell
instance ToNamedRecord Item where
  toNamedRecord Item{..} =
    Cassava.namedRecord
      [ "Item" .= itemName
      , "Link" .= itemLink
      , "Type" .= itemType
      ]
```

this and the corresponding `FromNamedRecord` both contain the header information in redundant form

the type of `toNamedRecord` is `toNamedRecord :: a -> NamedRecord`

To get the Field titles from `ToNamedRecord` you need an instance of the item..., similarly with `FromNamedRecord` you need
to actually parse something.... You can't extract this information without running the functions. This information however should
be accessible without text to parse and actual records to serialise... A csv containing no records and just a header line is a perfectly valid thing for some people to expect. (its probably just as reasonable to expect a CSV with no recoords to expect no header as well) the point is its fairly undefined and some people will expect one and others will expect the other...

The story is similar with `To/FromField`

Its for this reason that i would recommend in these cases not to bother with `ToNamedRecord` and just write the header yourself with conventional means. If you were writing a type that you were going to serialise into CSV, in addition to writing your `*NamedRecord` instances I would also recommend as a practice to also export your header text for this reason to your users, its outside of the `cassava` interface entirely but it just makes their lives easier for the reasons i described above:

``` haskell
headerLine :: [T.Text]
headerLine = ["Col1", "Col2"]
```
its just a list so they can choose their own separators

but basically it shouldn't require anything to get that info out, it should just be there...

I'd also highly recommend just getting familiar with plain IO and writing text to files in `IO` first before tackling things like `Conduit`. In many cases its more straight forward unless you have hard requirements on a smaller memory footprint

**Comment by mcgizzle** [comment]
Thanks for the info, thats really useful, I didn't realise I just needed to `sinkHandle` instead of `sinkFike`!

Regarding the `Header` stuff I don't really mind if its not from `NamedRecord`, when I was looking at casava API I noticed the header stuff felt a bit janky.

I'm streaming large files from bigQuery so there's a hard requirement on `Conduit`.

Thanks for your help, I'll close this ticket for now as I don't see the need anymore.
